File: Data Structures & Algorithms/lowest-common-ancestor-in-binary-search-tree/submission-0.py
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right

class Solution:
    def lowestCommonAncestor(self, root: TreeNode, p: TreeNode, q: TreeNode) -> TreeNode:
        curr = root

        while curr:
            # If both are greater than the current value,
            # then search in the right side.
            if p.val > curr.val and q.val > curr.val:
                curr = curr.right
            # If both are less than the current value,
            # then search in the left side.
            elif p.val < curr.val and q.val < curr.val:
                curr = curr.left
            else:
            # Return the current node.
                return curr

        return None

File: Data Structures & Algorithms/level-order-traversal-of-binary-tree/submission-0.cpp
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode() : val(0), left(nullptr), right(nullptr) {}
 *     TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
 *     TreeNode(int x, TreeNode *left, TreeNode *right) : val(x), left(left), right(right) {}
 * };
 */

class Solution {
private:
    vector<vector<int>> levels;

    void dfs(TreeNode* node, int depth) {
        if (node == NULL) { return; }

        if (levels.size() == depth) {
            levels.push_back(vector<int>());
        }

        levels[depth].push_back(node->val);
        dfs(node->left, depth + 1);
        dfs(node->right, depth + 1);
    }

public:
    vector<vector<int>> levelOrder(TreeNode* root) {
        if (root == NULL) { return {}; }

        dfs(root, 0);

        return levels;
    }
};
